### DISCUSSION TITLE
does renet work on real remote servers? deployment on ec2 doesn't work, maybe a bug?

### DISCUSSION THREAD

**Discussion by bcolloran** [lucaspoffo/renet#52] [Q&A] [answered]
Hi, sorry for the annoying beginner question, but I've been stuck on this for a couple days and I'm not making any progress...

I'm trying to get the [renet echo example](https://github.com/lucaspoffo/renet/blob/master/renet/examples/echo.rs) to work on ec2. I've verified that my ports are opened correctly and that I'm able to connect to my remote instance via UDP by following the instructions here: https://serverfault.com/a/678635. 

I thought that I'd just need to change the server from binding to localhost to an unspecified addess (i.e. bind the server to `"0.0.0.0:{}"` instead of `"127.0.0.1:{}"`, and allow the client to specify the IP of the ec2 server and to bind to an unspecified address as well.

Here's a gist of what I tried: https://gist.github.com/bcolloran/f499cfffcce3d8346f281a463934ef6a

When I start the process on the server with e.g. `echo_ec2 server 3412` and then run it on my local machine with `echo_ec2 client {IP of ec2 instance} 3412`, the local process times out after a bit with a timeout error:
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Netcode(Disconnected(ConnectionRequestTimedOut))', renet/examples/echo_ec2.rs:130:43
```

Any suggestions about what I might be doing wrong? Thanks so much!

**Comment by bcolloran** [comment]
I've been continuing to play with this for a few days, and I think there might be an issue with how renet binds to ports or connects to remote machines.

I've been able to get the Tokio  echo udp example ([client](https://github.com/tokio-rs/tokio/blob/master/examples/connect.rs), [server](https://github.com/tokio-rs/tokio/blob/master/examples/echo-udp.rs)) working on ec2 without any issue by binding both the client and server to `0.0.0.0:0` and then pointing the client at the public ip:

client command:
`connect --udp {public instance ip}:7777`

server:
`echo-udp 0.0.0.0:7777`


This works without any issues, successfully echoing input from the client to the server and back.

My gist from the post above takes the same approach: it binds sockets to `0.0.0.0:0` on both client and server, and then points the client at the public ip of the server (with the specified port). However, with renet it doesn't work. This suggests perhaps an issue with port binding or connection handling. Any ideas @lucaspoffo? Thank you!

**Reply by lucaspoffo** [reply, marked as answer]
Hey, sorry about the late reply

The problem seems to be at this line: https://gist.github.com/bcolloran/f499cfffcce3d8346f281a463934ef6a#file-echo_ec2-rs-L66

When creating the ServerConfig the server_address given to it must be the public address that the client will communicate to the server. So you can bind the socket to `0.0.0.0:0`, but you need to pass to this configuration the public address of the ec2 machine with the used port.

Edit: make sure you have port forwarding and the firewall (if the machine has one) working correctly. This is a common problem.

**Reply by 0crispy** [reply]
Even with the socket bound to the public ip the client still isn't able to connect to the server. I get the same error (Netcode(Disconnected(ConnectionRequestTimedOut))).

Edit: It was just a port-forward issue.
